src/dodal/common/beamlines/instantiation_behaviour.py

```diff
@@ -0,0 +1,131 @@
+from collections.abc import Callable
+from dataclasses import dataclass
+
+from bluesky.run_engine import call_in_bluesky_event_loop
+from ophyd_async.core import DEFAULT_TIMEOUT
+from ophyd_async.core import Device as OphydV2Device
+
+from dodal.common.beamlines.beamline_utils import (
+    ACTIVE_DEVICES,
+)
+from dodal.utils import AnyDevice
+
+
+class XYZDetector(OphydV2Device):
+    def __init__(self, prefix: str, *args, **kwargs):
+        self.prefix = prefix
+        super().__init__(*args, **kwargs)
+
+    @property
+    def hints(self):
+        raise NotImplementedError
+
+
+@dataclass
+class DeviceInitializationConfig:
+    """
+    eager ones all connect at the beginning in blueapi startup. if they fail to connect we get an error
+    endpoint for what devices are failing
+    only matters in connect failure reporting behavior
+    lazy and eager are not about when connect but whether should be connected
+    """
+
+    eager: bool = True
+    set_name: bool = True
+    default_timeout_for_connect: float = DEFAULT_TIMEOUT
+    default_use_mock_at_connection: bool = False
+
+
+class DeviceInitializationController:
+    """
+    This class is responsible for the instantiation of a device.
+    The device is then cached here locally, and also in the global variable ACTIVE_DEVICES.
+    """
+
+    device: AnyDevice | None = None
+    factory: Callable[[], AnyDevice]
+
+    def __init__(
+        self, config: DeviceInitializationConfig, factory: Callable[[], AnyDevice]
+    ) -> None:
+        self.factory = factory
+        self.config = config
+        CONTROLLERS[self.factory.__name__] = self
+        self.device = None
+
+    def see_if_device_is_in_cache(self) -> AnyDevice | None:
```

Comment:
Should: I would prefer more conventional terminology here:

```suggestion
    def get_device_if_cached(self) -> AnyDevice | None:
```

Reply:
will implement suggestions tomorrow

Reply:
done